File: backend/app/data_sources/options_data.py
"""Options chain data and Greeks from Massive API."""

import logging
import numpy as np
from datetime import datetime
from app.data_sources import massive

logger = logging.getLogger(__name__)


def fetch_options_data(tickers: list[str]) -> dict:
    """Fetch ATM options with Greeks for each ticker."""
    results = {}
    for ticker_sym in tickers:
        try:
            snap = massive.fetch_snapshot(ticker_sym)
            current_price = snap.get("current_price") if snap else None
            if not current_price:
                results[ticker_sym] = {"has_options": False, "ticker": ticker_sym, "error": "No price"}
                continue

            chain_data = massive.fetch_options_chain(ticker_sym)
            if not chain_data.get("has_options"):
                results[ticker_sym] = {"has_options": False, "ticker": ticker_sym}
                continue

            contracts = chain_data.get("contracts", [])
            if not contracts:
                results[ticker_sym] = {"has_options": False, "ticker": ticker_sym}
                continue

            # Group by expiration, find nearest 7-45 days out
            expirations = set()
            for c in contracts:
                exp = c.get("expiration_date")
                if exp:
                    expirations.add(exp)

            expiry = _find_nearest_expiration(sorted(expirations))
            if not expiry:
                results[ticker_sym] = {"has_options": False, "ticker": ticker_sym}
                continue

            # Filter contracts for this expiration
            exp_contracts = [c for c in contracts if c.get("expiration_date") == expiry]
            calls = [c for c in exp_contracts if c.get("contract_type") == "call"]
            puts = [c for c in exp_contracts if c.get("contract_type") == "put"]

            atm = _find_atm_options(calls, puts, current_price)

            # Historical volatility
            hv = _compute_historical_volatility(ticker_sym)

            # IV from ATM options
            iv_call = atm.get("call", {}).get("implied_volatility")
            iv_put = atm.get("put", {}).get("implied_volatility")
            iv_avg = None
            if iv_call is not None and iv_put is not None:
                iv_avg = round((iv_call + iv_put) / 2, 4)
            elif iv_call is not None:
                iv_avg = iv_call
            elif iv_put is not None:
                iv_avg = iv_put

            vol_comparison = _compare_iv_hv(iv_avg, hv)

            exp_date = datetime.strptime(expiry, "%Y-%m-%d") if len(expiry) == 10 else datetime.fromisoformat(expiry)
            days_to_expiry = (exp_date - datetime.now()).days

            results[ticker_sym] = {
                "ticker": ticker_sym,
                "has_options": True,
                "expiration": expiry[:10],
                "days_to_expiry": max(days_to_expiry, 0),
                "atm_strike": atm.get("strike"),
                "call": atm.get("call", {}),
                "put": atm.get("put", {}),
                "volatility": {
                    "iv_call": iv_call,
                    "iv_put": iv_put,
                    "iv_avg": iv_avg,
                    "hv_30d": hv,
                    "iv_hv_ratio": round(iv_avg / hv, 2) if iv_avg and hv and hv > 0 else None,
                    **vol_comparison,
                },
            }
        except Exception as e:
            results[ticker_sym] = {"has_options": False, "ticker": ticker_sym, "error": str(e)}
    return results


def _find_nearest_expiration(expirations: list[str], min_days: int = 7, max_days: int = 45) -> str | None:
    """Find nearest expiration in the 7-45 day window."""
    now = datetime.now()
    best = None
    best_diff = float("inf")

    for exp_str in expirations:
        try:
            exp_date = datetime.strptime(exp_str[:10], "%Y-%m-%d")
        except ValueError:
            continue
        days = (exp_date - now).days
        if min_days <= days <= max_days and days < best_diff:
            best = exp_str
            best_diff = days

    if best is None:
        for exp_str in expirations:
            try:
                exp_date = datetime.strptime(exp_str[:10], "%Y-%m-%d")
            except ValueError:
                continue
            days = (exp_date - now).days
            if days > 0 and days < best_diff:
                best = exp_str
                best_diff = days

    return best


def _find_atm_options(calls: list[dict], puts: list[dict], current_price: float) -> dict:
    """Find nearest ATM call and put."""
    if not calls and not puts:
        return {"strike": None, "call": {}, "put": {}}

    all_strikes = [c.get("strike_price", 0) for c in calls] + [p.get("strike_price", 0) for p in puts]
    if not all_strikes:
        return {"strike": None, "call": {}, "put": {}}

    strikes = np.array(all_strikes)
    atm_strike = float(strikes[np.argmin(np.abs(strikes - current_price))])

    call_row = next((c for c in calls if c.get("strike_price") == atm_strike), None)
    put_row = next((p for p in puts if p.get("strike_price") == atm_strike), None)

    def _extract(row) -> dict:
        if row is None:
            return {}
        return {
            "strike": row.get("strike_price"),
            "last_price": _safe(row.get("last_price")),
            "bid": _safe(row.get("bid")),
            "ask": _safe(row.get("ask")),
            "implied_volatility": _safe(row.get("implied_volatility")),
            "open_interest": int(row.get("open_interest") or 0),
            "volume": int(row.get("volume") or 0),
            "delta": _safe(row.get("delta")),
            "gamma": _safe(row.get("gamma")),
            "theta": _safe(row.get("theta")),
            "vega": _safe(row.get("vega")),
        }

    return {
        "strike": atm_strike,
        "call": _extract(call_row),
        "put": _extract(put_row),
    }


def _compute_historical_volatility(ticker_sym: str, period: int = 30) -> float | None:
    """Compute 30-day annualized historical volatility."""
    try:
        import pandas as pd
        history = massive.fetch_history(ticker_sym, period="3mo")
        if len(history) < period:
            return None
        closes = pd.Series([h["close"] for h in history])
        log_returns = np.log(closes / closes.shift(1)).dropna()
        hv = float(log_returns.tail(period).std() * np.sqrt(252))
        return round(hv, 4)
    except Exception:
        return None


def _compare_iv_hv(iv: float | None, hv: float | None) -> dict:
    """Compare IV vs HV and flag opportunities."""
    if iv is None or hv is None or hv == 0:
        return {"signal": "N/A", "opportunity": "Insufficient volatility data"}

    ratio = iv / hv
    if ratio > 1.2:
        return {
            "signal": "IV_ELEVATED",
            "opportunity": "IV exceeds HV — options are expensive. Consider selling covered calls or credit spreads.",
        }
    elif ratio < 0.8:
        return {
            "signal": "IV_DEPRESSED",
            "opportunity": "IV below HV — options are cheap. Consider buying protective puts or debit spreads.",
        }
    return {
        "signal": "IV_NORMAL",
        "opportunity": "IV roughly in line with HV — no clear volatility edge.",
    }


def _safe(val) -> float | None:
    if val is None:
        return None
    try:
        f = float(val)
        return None if np.isnan(f) else round(f, 4)
    except (TypeError, ValueError):
        return None

File: laravel-backend/app/DataSources/OptionsData.php
<?php

namespace App\DataSources;

use Carbon\Carbon;
use Illuminate\Support\Facades\Log;

/**
 * Options chain data and Greeks from Massive.com REST API.
 */
class OptionsData
{
    private MassiveClient $massiveClient;
    private AlphaVantage $alphaVantage;
    private YahooFinance $yahooFinance;

    public function __construct(MassiveClient $massiveClient, AlphaVantage $alphaVantage, YahooFinance $yahooFinance)
    {
        $this->massiveClient = $massiveClient;
        $this->alphaVantage = $alphaVantage;
        $this->yahooFinance = $yahooFinance;
    }

    public function fetchOptionsData(array $tickers): array
    {
        $results = [];

        foreach ($tickers as $ticker) {
            try {
                $avQuote = $this->alphaVantage->fetchQuote($ticker);
                $currentPrice = $avQuote['current_price'] ?? null;

                if (! $currentPrice) {
                    $results[$ticker] = ['has_options' => false, 'ticker' => $ticker, 'error' => 'No price'];
                    continue;
                }

                $minDate = Carbon::now()->addDays(7)->format('Y-m-d');
                $maxDate = Carbon::now()->addDays(45)->format('Y-m-d');

                $options = $this->massiveClient->getOptionsChain($ticker, [
                    'expiration_date.gte' => $minDate,
                    'expiration_date.lte' => $maxDate,
                    'strike_price.gte' => $currentPrice * 0.95,
                    'strike_price.lte' => $currentPrice * 1.05,
                ]);

                if (empty($options)) {
                    $options = $this->massiveClient->getOptionsChain($ticker, [
                        'expiration_date.gte' => $minDate,
                        'strike_price.gte' => $currentPrice * 0.90,
                        'strike_price.lte' => $currentPrice * 1.10,
                    ]);
                }

                if (empty($options)) {
                    $results[$ticker] = ['has_options' => false, 'ticker' => $ticker];
                    continue;
                }

                $calls = array_filter($options, fn($o) =>
                    ($o['details']['contract_type'] ?? '') === 'call'
                );
                $puts = array_filter($options, fn($o) =>
                    ($o['details']['contract_type'] ?? '') === 'put'
                );

                if (empty($calls) && empty($puts)) {
                    $results[$ticker] = ['has_options' => false, 'ticker' => $ticker];
                    continue;
                }

                $atmCall = $this->findAtm(array_values($calls), $currentPrice);
                $atmPut = $this->findAtm(array_values($puts), $currentPrice);

                $atmOption = $atmCall ?? $atmPut;
                $strike = $atmOption['details']['strike_price'] ?? null;
                $expiry = $atmOption['details']['expiration_date'] ?? null;
                $daysToExpiry = $expiry
                    ? max(0, Carbon::parse($expiry)->diffInDays(Carbon::now()))
                    : 0;

                $callData = $atmCall ? $this->extractOption($atmCall) : [];
                $putData = $atmPut ? $this->extractOption($atmPut) : [];

                $ivCall = $callData['implied_volatility'] ?? null;
                $ivPut = $putData['implied_volatility'] ?? null;
                $ivAvg = null;
                if ($ivCall !== null && $ivPut !== null) {
                    $ivAvg = round(($ivCall + $ivPut) / 2, 4);
                } elseif ($ivCall !== null) {
                    $ivAvg = $ivCall;
                } elseif ($ivPut !== null) {
                    $ivAvg = $ivPut;
                }

                $hv = $this->computeHistoricalVolatility($ticker);
                $volComparison = $this->compareIvHv($ivAvg, $hv);

                $results[$ticker] = [
                    'ticker' => $ticker,
                    'has_options' => true,
                    'expiration' => $expiry,
                    'days_to_expiry' => $daysToExpiry,
                    'atm_strike' => $strike,
                    'call' => $callData,
                    'put' => $putData,
                    'volatility' => array_merge([
                        'iv_call' => $ivCall,
                        'iv_put' => $ivPut,
                        'iv_avg' => $ivAvg,
                        'hv_30d' => $hv,
                        'iv_hv_ratio' => ($ivAvg && $hv && $hv > 0) ? round($ivAvg / $hv, 2) : null,
                    ], $volComparison),
                ];
            } catch (\Throwable $e) {
                Log::error("Options data failed for {$ticker}: {$e->getMessage()}");
                $results[$ticker] = ['has_options' => false, 'ticker' => $ticker, 'error' => $e->getMessage()];
            }
        }

        return $results;
    }

    private function findAtm(array $options, float $currentPrice): ?array
    {
        if (empty($options)) {
            return null;
        }

        return collect($options)->sortBy(function ($o) use ($currentPrice) {
            return abs(($o['details']['strike_price'] ?? 0) - $currentPrice);
        })->first();
    }

    private function extractOption(array $option): array
    {
        $result = [];
        $result['strike'] = $option['details']['strike_price'] ?? null;
        $result['last_price'] = $option['day']['close'] ?? null;
        $result['bid'] = $option['last_quote']['bid'] ?? null;
        $result['ask'] = $option['last_quote']['ask'] ?? null;
        $result['implied_volatility'] = $this->safe($option['implied_volatility'] ?? null);
        $result['open_interest'] = (int) ($option['open_interest'] ?? 0);
        $result['volume'] = (int) ($option['day']['volume'] ?? 0);

        $greeks = $option['greeks'] ?? [];
        $result['delta'] = $this->safe($greeks['delta'] ?? null);
        $result['gamma'] = $this->safe($greeks['gamma'] ?? null);
        $result['theta'] = $this->safe($greeks['theta'] ?? null);
        $result['vega'] = $this->safe($greeks['vega'] ?? null);

        return $result;
    }

    private function computeHistoricalVolatility(string $ticker, int $period = 30): ?float
    {
        try {
            $history = $this->yahooFinance->fetchHistory($ticker, '3mo');
            if (count($history) < $period) {
                return null;
            }

            $closes = array_column($history, 'close');
            $logReturns = [];
            for ($i = 1; $i < count($closes); $i++) {
                if ($closes[$i - 1] > 0) {
                    $logReturns[] = log($closes[$i] / $closes[$i - 1]);
                }
            }

            $recent = array_slice($logReturns, -$period);
            if (count($recent) < 2) {
                return null;
            }

            $mean = array_sum($recent) / count($recent);
            $sumSquares = array_sum(array_map(fn($x) => ($x - $mean) ** 2, $recent));
            $std = sqrt($sumSquares / (count($recent) - 1));
            $hv = $std * sqrt(252);

            return round($hv, 4);
        } catch (\Throwable $e) {
            return null;
        }
    }

    private function compareIvHv(?float $iv, ?float $hv): array
    {
        if ($iv === null || $hv === null || $hv == 0) {
            return ['signal' => 'N/A', 'opportunity' => 'Insufficient volatility data'];
        }

        $ratio = $iv / $hv;
        if ($ratio > 1.2) {
            return [
                'signal' => 'IV_ELEVATED',
                'opportunity' => 'IV exceeds HV — options are expensive. Consider selling covered calls or credit spreads.',
            ];
        } elseif ($ratio < 0.8) {
            return [
                'signal' => 'IV_DEPRESSED',
                'opportunity' => 'IV below HV — options are cheap. Consider buying protective puts or debit spreads.',
            ];
        }

        return [
            'signal' => 'IV_NORMAL',
            'opportunity' => 'IV roughly in line with HV — no clear volatility edge.',
        ];
    }

    private function safe(mixed $val): ?float
    {
        if ($val === null) {
            return null;
        }
        $f = (float) $val;
        return is_nan($f) ? null : round($f, 4);
    }
}
